staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -356,3 +365,62 @@ func shouldRecordWatchListLatency(event watch.Event) bool {
 	}
 	return hasAnnotation
 }
+
+// computeEventTransformerFunction returns a transformer function for watch events.
+//
+// note that: for simplicity, nil emptyVersionedList indicates a standard watch request for
+// which we return a no-op transformer.
+// otherwise, we assume a watchlist request for which we return a transformer
+// that embeds the given list.
+func computeEventTransformerFunction(emptyVersionedList runtime.Object) func(watch.Event, runtime.Encoder) watch.Event {
+	if emptyVersionedList == nil {
+		return func(e watch.Event, _ runtime.Encoder) watch.Event {
+			return e
+		}
+	}
+	return func(event watch.Event, encoder runtime.Encoder) watch.Event {
+		return annotateInitialEventsEmbeddedListBookmark(event, emptyVersionedList, encoder)
+	}
+}
+
+// annotateInitialEventsEmbeddedListBookmark adds a special annotation to the given object
+// which contains an empty base64 encoded list that is used by client to construct the final response.
+func annotateInitialEventsEmbeddedListBookmark(event watch.Event, emptyVersionedList runtime.Object, encoder runtime.Encoder) watch.Event {
+	if event.Type != watch.Bookmark {
+		return event
+	}
+	if objectMeta, err := meta.Accessor(event.Object); err != nil {
+		return newWatchErrorFor(err)
+	} else if objectMeta.GetAnnotations()[metav1.InitialEventsAnnotationKey] != "true" {
+		return event
+	}
+
+	event.Object = event.Object.DeepCopyObject()
+	objectMeta, err := meta.Accessor(event.Object)
+	if err != nil {
+		return newWatchErrorFor(err)
+	}
+	annotations := objectMeta.GetAnnotations()
+	if annotations == nil {
```

Comment:
nit: this cannot happen right? you already checked above that InitialEventsAnnotationKey is true :)